doc/convergence_curves.rst

```diff
@@ -0,0 +1,51 @@
+.. _convergence_curves:
+
+Sampling of the convergence curves
+==================================
+
+For each solver, there are two ways to create a convergence curve.
+They are chosen by the ``stopping_strategy`` attribute of the solver.
+
+1. Using iterations or tolerance
+--------------------------------
+
+
+The first way is to use ``Solver.stopping_strategy = "iteration"`` or ``Solver.stopping_strategy = "tolerance"``.
+This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or numerical tolerance.
+This stopping strategy creates curves by calling ``Solver.run(stop_val)`` several times with different values for the ``stop_val`` parameter:
+
+- if the solver's ``stopping_strategy`` is ``"iteration"``, ``stop_val`` is  the number of iterations passed to ``run``.
+  It increases geometrically by at least 1, starting from 1 with a factor 1.5.
+  Note that the first call uses a number of iterations of 0.
+
+- if the solver's ``stopping_strategy`` is ``"tolerance"``, the ``stop_val`` parameter corresponds to the numerical tolerance.
+  It decreases geometrically by a factor of 1.5 between each call to ``run``, starting from 1 at the second call.
+  Note that the first call uses a tolerance of 1e38.
+
+In both cases, if the objective curve is flat (i.e., the variation of the objective between two points is numerically 0), the geometric rate is multiplied by 1.2.
+
+Note that the solver is started from scratch at each call to ``solver.run``.
+For more advanced configurations, the evolution of ``stop_val`` can be controlled on a per solver basis, by implementing a static  ``Solver.get_next`` method, which receives the current value for tolerance/number of iterations, and returns the next one.
+
+2. Using a callback
+-------------------
+
+restarting the solver from scratch, though inevitable to handle black box solvers, can be costly.
+
+When a solver exposes the intermediate values of the iterates, it is possible to create the curve in a single solver run, by using ``stopping_strategy = "callback"``.
+In that case, the argument passed to ``Solver.run`` will be a callable object, ``callback``.
+Every time it is called inside ``Solver.run`` with the current iterate ``x`` as argument, this object will store the objective value at the current iterate, and the current running time.
```

Comment:
It is not this.
Actually, `stopping_strategy == "callback"` follows the same principle as `"iteration"`, except that `n_iter` is replaced by the number of call to the callback.

The loss is computed only when reaching this `stop_val` iterations and `shoud_stop` is also only called at this moment.

I think I agree with @jeandut that this is a bit a "melange des genres". I am all up to improve this, but the best would be to take time to discuss what we wanna do, to have a proper solution. Happy to chat IRL with you two.

Reply:
I don't get it, there is a single loss computed per run, which is the final one?

Reply:
No, the loss is computed at the same point where it would have been with `stopping_strategy == "iteration"`, and the `should_stop` is also evaluated at the same points.
Basically, `callback` is a mirror for `iteration`, where the number of iterations is equal to the number of call to the callback.

Reply:
A small pass on this text.
```suggestion
Like with ``stopping_strategy == "iteration"``, the loss is computed after a number of callback's calls that grows geometrically. If the loss was computed after :math:`n` calls, the loss and timing will be computed again when reaching :math:`\max(n+1, \rho * n)` calls to the callback.
The callback makes sure we do not account for loss computation time and also check for convergence every time the loss is computed (as described in the next section).
It returns ``False`` when the solver should be stopped. A classical usage pattern is:

.. code:: python

    def run(self, callback):
        x = ... # Initialize iterate

        while callback(x):
            x = ...  # Update iterate
        self.x = x

```